otlp/traces_direct_json.go

```diff
@@ -0,0 +1,1115 @@
+	ctx context.Context,
+	v *fastjson.Value,
+	traceID,
+	parentSpanID,
+	parentName []byte,
+	spanStartTime uint64,
+	resourceAttrs,
+	scopeAttrs *msgpAttributes,
+	sampleRate int32,
+	isError bool,
+	batch *BatchMsgp,
+) (*msgpAttributes, error) {
+	eventAttr := msgpAttributesPool.Get().(*msgpAttributes)
+
+	// Set trace info
+	if len(traceID) > 0 {
+		eventAttr.addTraceID([]byte("trace.trace_id"), traceID)
+	}
+	if len(parentSpanID) > 0 {
+		eventAttr.addHexID([]byte("trace.parent_id"), parentSpanID)
+	}
+
+	// Mark as span event
+	eventAttr.addString([]byte("meta.annotation_type"), []byte("span_event"))
+	eventAttr.addString([]byte("meta.signal_type"), []byte("trace"))
+	if len(parentName) > 0 {
+		eventAttr.addString([]byte("parent_name"), parentName)
+	}
+
+	// Parse event fields
+	var name []byte
+	var timeUnixNano uint64
+	var exceptionAttrs *msgpAttributes
+	var attributes []*fastjson.Value
+
+	obj, err := v.Object()
+	if err != nil {
+		return nil, err
+	}
+
+	obj.Visit(func(key []byte, v *fastjson.Value) {
+		switch string(key) {
+		case "timeUnixNano":
+			// Time is encoded as string in JSON
+			strBytes := v.GetStringBytes()
+			var val uint64
+			val, err = strconv.ParseUint(string(strBytes), 10, 64)
+			if err == nil {
+				timeUnixNano = val
+			}
+
+		case "name":
+			name = v.GetStringBytes()
+
+		case "attributes":
+			attributes = v.GetArray()
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(attributes) > 0 {
+		if err = unmarshalKeyValueArrayJSON(ctx, attributes, eventAttr, 0); err != nil {
+			return nil, err
+		}
+
+		// If this is an exception event, also collect exception-specific attributes
+		if bytes.Equal(name, []byte("exception")) {
+			exceptionAttrs = msgpAttributesPool.Get().(*msgpAttributes)
+
+			// Extract exception attributes
+			for _, kv := range attributes {
+				kvObj, err := kv.Object()
+				if err != nil {
+					continue
+				}
+
+				var keyStr string
+				var valueObj *fastjson.Value
+
+				kvObj.Visit(func(k []byte, v *fastjson.Value) {
+					switch string(k) {
+					case "key":
+						keyStr = string(v.GetStringBytes())
+					case "value":
+						valueObj = v
+					}
+				})
+
+				if keyStr == "" || valueObj == nil {
+					continue
+				}
+
+				switch keyStr {
+				case "exception.message", "exception.type", "exception.stacktrace":
+					valObj, err := valueObj.Object()
+					if err == nil {
+						valObj.Visit(func(k []byte, v *fastjson.Value) {
+							if string(k) == "stringValue" {
+								valBytes := v.GetStringBytes()
+								exceptionAttrs.addString([]byte(keyStr), valBytes)
+							}
+						})
+					}
+				case "exception.escaped":
+					valObj, err := valueObj.Object()
+					if err == nil {
+						valObj.Visit(func(k []byte, v *fastjson.Value) {
+							if string(k) == "boolValue" {
+								exceptionAttrs.addBool([]byte(keyStr), v.GetBool())
+							}
+						})
+					}
+				}
+			}
+		}
+	}
+
+	// Set event name
+	eventAttr.addString([]byte("name"), name)
```

Comment:
It would be nice to share the same logic for how we handle span event between JSON and protobuf implementation